Add support for servercheck updates to go client

guzzijason opened the pull request:
## What does this PR (Pull Request) do?
Adding support for the `/api/1.3/servercheck` route to the go client, which was previously missing.

- This PR is not related to any Issue
- Documentation via godoc


## Which Traffic Control components are affected by this PR?

- Traffic Control Client (go)
- Traffic Ops servercheck scripts (WIP, new go version to replace legacy perl)

## What is the best way to verify this PR?

Client test code used:
```
        var statusData tc.ServercheckNullable
        statusData.ID = server.ID
        statusData.Name = "DSCP"
        statusData.Value = 0
        _, err := session.UpdateCheckStatus(statusData)
        if err != nil {
            panic(err)
        }
```
No tests were included because no API changes were made.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

asfgit commented:
Can one of the admins verify this patch?

dangogh commented:
add to whitelist

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3483/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3484/
Test FAILed.

dangogh commented:
sorry -- only had a cursory look at it before..   The code looks good,  but I'd like to see a test for it. It should be fairly straightforward:  in `traffic_ops/testing/api/v14`,  add a new `servercheck_test.go` and fill in json in `tc-fixtures.json`.   Simple test to make sure you can read the data when it goes into the table as well as any tests to check for incorrect input.  Plenty of other examples there.  I can help out with that if needed.

guzzijason commented:
I think I'll need to add a client function to read the data then - will probably be helpful for the test :)

rob05c commented on traffic_ops/client/servercheck.go:
```diff
@@ -0,0 +1,36 @@
+const API_V13_SERVERCHECK = "/api/1.3/servercheck"
+
+// Update a Server Check Status
+func (to *Session) UpdateCheckStatus(status tc.ServercheckNullable) (*tc.ServercheckPostResponse, error) {
```
Mind changing this to return `ReqInf`? 

Should just be 

`func (to *Session) UpdateCheckStatus(status tc.ServercheckNullable) (*tc.ServercheckPostResponse, ReqInf, error) {`

```
reqInf, err = post(to, uri, jsonReq, &resp)
if err != nil {
		return nil, reqInf, err
	}
	return &resp, reqInf, nil
```

We're working to change all client funcs to return `ReqInf`. It has data that helps clients debug and return better error messages.

guzzijason replied:
Sure, will do.

guzzijason commented:
So, I added a test, but I'm not 100% sure it's what is needed. Currently, the only use case is to update the check status, so it's pretty simplistic, I think. Also, I'm not sure the best way to handle the fact that the "extension" use is required to be used for the `servercheck` API route. I added that user to the test, but I doubt it will authenticate as-is.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3487/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3667/
Test PASSed.

ocket8888 commented:
The API test for this route doesn't compile, load the data it needs to run, or run as the proper user. It also causes a runtime panic when run.
